REGISTRY-14: Add a new module for spnego kerberos authorization

priyank5485 opened the pull request:
I have disabled 3 tests related to kerberos since they require kdc setup. Initial request or any other request made by the browser to ui server works fine. However, there is a need to change some ui code so that subsequent http requests made by ui to the server on user clicking different controls also passes the same cookie that was generated as result of authentication. Currently, subsequent requests ignore the cookie and hence fail to authenticate. It should be a quick fix in the UI.

harshach commented:
@priyank5485 Nice Work!!. We also need to handle this for SchemaRegistryClient as it uses REST API to query .

harshach commented:
overall looks good to me. It would good to have SECURITY.md talking about the necessary configs to setup the secure env.

satishd commented on schema-registry-auth/src/main/java/com/hortonworks/registries/auth/client/KerberosAuthenticator.java:
```diff
@@ -0,0 +1,370 @@
+ */
+public class KerberosAuthenticator implements Authenticator {
+
+    private static Logger LOG = LoggerFactory.getLogger(
```
nit: no new line required here.

priyank5485 replied:
Will fix.

satishd commented on schema-registry-auth/src/main/java/com/hortonworks/registries/auth/client/KerberosAuthenticator.java:
```diff
@@ -0,0 +1,370 @@
+                                KerberosAuthenticator.this.url.getHost());
+                        Oid oid = KerberosUtil.getOidInstance("NT_GSS_KRB5_PRINCIPAL");
+                        GSSName serviceName = gssManager.createName(servicePrincipal,
+                                oid);
```
minor nit: avoid new line here.

priyank5485 replied:
Will fix.

satishd commented on schema-registry-auth/src/main/java/com/hortonworks/registries/auth/client/KerberosAuthenticator.java:
```diff
@@ -0,0 +1,370 @@
+                                oid);
+                        oid = KerberosUtil.getOidInstance("GSS_KRB5_MECH_OID");
+                        gssContext = gssManager.createContext(serviceName, oid, null,
+                                GSSContext.DEFAULT_LIFETIME);
```
minor nit: avoid new line here.

priyank5485 replied:
Will fix.

satishd commented on schema-registry-auth/src/main/java/com/hortonworks/registries/auth/client/KerberosAuthenticator.java:
```diff
@@ -0,0 +1,370 @@
+                    } finally {
+                        if (gssContext != null) {
+                            gssContext.dispose();
+                            gssContext = null;
```
nit: no need to set this as null as it is created in this method.

priyank5485 replied:
Will fix.

satishd commented:
@priyank5485 It seems large part of the code is borrowed from `hadoop-auth` module. Why can not we use that module and extend wherever it is possible instead of borrowing all those sources into this repo? 

cc @harshach

HeartSaVioR commented:
AFAIK, if the code is borrowed from hadoop-auth, the original license (copyright) should be kept.
That's what Heron does for borrowing Apache Storm's codebase.

priyank5485 commented on schema-registry-auth/src/main/java/com/hortonworks/registries/auth/client/AuthenticatedURL.java:
```diff
@@ -0,0 +1,278 @@
+    /**
+     * Client side authentication token.
+     */
+    public static class Token {
```
@satishd I dont see a strong reason for using the Token inner class. Its mainly to avoid returning new strings(tokens) as part of some method calls in client for a connection made to an already authenticated url. As you mentioned the code has been borrowed from hadoop-auth module and thats how it was implemented.

priyank5485 commented on schema-registry-auth/src/main/java/com/hortonworks/registries/auth/client/AuthenticationException.java:
```diff
@@ -0,0 +1,52 @@
+package com.hortonworks.registries.auth.client;
+
+/**
+ * Exception thrown when an authentication error occurrs.
```
Will fix.

priyank5485 commented:
@harshach This module also has classes for non browser clients. I think I can incorporate them with schema registry client. Will update the PR. I will also add SECURITY.md file with it.

@satishd Main reason to create a module of our own is to not bring in some other dependencies like zookeeper or curator and get rid of code/classes we don't need. Even though classes can be extended and modified to change behavior to suit our needs, this gives us more control regarding changing code however we want to.

@HeartSaVioR Yes, all this code is borrowed from hadoop-auth module and changed slightly based on our needs. I have removed the Copyright Hortonworks part from the license which makes the license pretty much how it was on the original hadoop-auth files. Please let me know if this looks good or not

harshach commented:
@satishd @priyank5485 @HeartSaVioR we did the same thing in storm-core by making a dependency on hadoop-auth. It's turning into a PITA as we have topologies who depend on their own version of hadoop which also brings a different version of hadoop-auth. Yahoo! guys ran into issues they also tried removed this dependency from storm-core. I want to avoid potential issues because of that. It also makes it easy to change and improve as per our needs.

harshach commented:
Licensing info looks good to me.

HeartSaVioR commented:
@priyank5485 
https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-auth/src/main/java/org/apache/hadoop/util/PlatformName.java

Origin file clearly states that license is belong to ASF, not only describing that that is ASL v2.0. Let's keep the **origin license comment** and don't modify that.

satishd commented:
@priyank5485 @harshach @HeartSaVioR It is OK to borrow if the dependency issues are getting complicated with user code etc. But we will also be missing any fixes that go into `hadoop-auth` module which we have to keep an eye on that module before every streamline release and update these sources with `apache/hadoop-auth` fixes. 

 It may be good to document the reason of borrowing in README.md of this module.

harshach commented:
@satishd Agree on documenting the details and definitely we should keep track of changes going into hadoop-auth's SPNEGO.

hmcl commented:
@harshach @priyank5485 @HeartSaVioR @satishd I read your discussion about the `hadoop-auth` dependencies. There is no easy answer here attending to the whole dependencies discussion, but duplicating the source code will make it very difficult to keep track of any fixes that go into `hadoop-auth` on a timely manner. 

We also need similar code in Streamline. Should we duplicate the code in Streamline as well ?

I would suggest that If depending on `hadoop-auth` is really an impossibility, perhaps we should have a streamline/registry maven repo just with this code that we could maintain, and simply add a dependency in maven on both streamline and registry.

satishd commented:
@hmcl I guess the plan is to use this module in `streamline` by having a dependency on this auth module. @harshach may want to comment on that.

@priyank5485 new module for auth should be `common-auth` or `registry-auth` or some other relevant name instead of `schema-registry-auth` as this module is more of generic authentication and not really only about schema-registry. Schema registry related stuff can be pushed into a submodule in schema-registry.

satishd commented on schema-registry-auth/pom.xml:
```diff
@@ -0,0 +1,134 @@
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>schema-registry-auth</artifactId>
```
@priyank5485 You may want to rename this module as commented [here](https://github.com/hortonworks/registry/pull/52#issuecomment-288287062)

HeartSaVioR commented:
@hmcl 
I guess you're suggesting having common project. That would make sense to me, since Registry is based on Streamline, and shares lots of codebase.

We're taking two different approach simultaneously - copy and modify codes to make in sync, and referring other project's module - and neither is ideal for me. IMHO, Streamline just needs to refer Registry for provided client, not refer common module. (it only refers `Schema` class indeed...) If we have versions for two projects, it will be more complicated.

What @hmcl suggests are promising, but I'm OK to duplicate the codebase if it feels too ideal for us.

cc. @harshach @priyank5485 @satishd

satishd commented:
@HeartSaVioR registry is not based or dependent on streamline but it is the other way. I have put my opinion in my earlier [comment](https://github.com/hortonworks/registry/pull/52#issuecomment-288287062)

satishd commented:
@HeartSaVioR Let me elaborate on my earlier comment here. 
What I mean here is registry does not really depend upon any streamline module. We wanted storage component to be in registry as it is kind of individual component shared by both streamline and registry. 

It is ideal to have one more project which contains common stuff but it adds more logistics. It may be practical to have the common modules in streamline or registry and depend on respective modules.

cc @harshach @priyank5485 @hmcl

hmcl commented:
@HeartSaVioR @satishd @priyank5485 @harshach in essence all I mean is that many projects need an authentication (and security - for that matter) module just like streamline and registry now do. Currently Hadoop has that in it's own **hadoop-auth** module. The natural thing for me would be to have a project called **cluster-auth** or **cluster-security** or something along those lines, and refer to this project using maven coordinates. This project would have what currently exists in **hadoop-auth**.  In the limit, even Hadoop itself could depend on this project in the future (perhaps this is a bit too ambitious, but several projects started as a component of a project and became their own entity). 

This would be the ideal world scenario for me. In the short time frame maybe this is not doable. However, we can give some strides in this direction by putting all the necessary code in a security module in registry (since streamline depends on it), and have streamline refer to it. Perhaps in the future we can extract it to become it's own project. In essence, make it modular, and if we have to duplicate it for now, let's put it all in one place, and such a way that we can extract it at any point in the future.

For now I am OK with having this in registry and I will refer to it in streamline if referring to hadoop-auth is really really unpractical.

harshach commented:
@satishd @priyank5485 @hmcl @HeartSaVioR 
Agree with most points made here. But SPNEGO auth is only useful for UI based auth and it doesn't solve more complext Kerberos auth for services etc.. . Hence our initial idea was to port this into registry and build that as a independent module and use that in streamline.
I think points made here are aligns with this and Instead of making this as a independent project, for now lets make this as a module under registry. Streamline already depends on registry modules and we can use this module in Streamline.

@priyank5485 lets see if can add another module to registry under com.hortonworks.spengo-auth and not under registries. So that we can keep it as independent as it can be.
We are already publishing registry artifacts into maven central we will do the same with spengo-auth.

harshach commented:
@priyank5485 @satishd @HeartSaVioR @hmcl 

Just to make it clear the intentions here.

1. Registry is an independent project and doesn't need any dependencies from streamline.
2. Streamline depends on Registry for schemas and other things.
3. Registry & Streamline shares same storage & common part.

What are we trying to do:
1. Remove common code from Streamline repository
2. Move any modules that are appropriate to be in Registry like MLRegistry .
3. Make Streamline depend on registry storage & common modules
4. Use hadoop-auth SPNEGO code in registry module with modifications. (Reasons are already provided in above comments).
5. Make streamline depend on registry authentication module.

I hope the above clears up any confusion. There isn't any code duplication going on instead we are removing the current code duplication.

I hear your concerns about copying hadoop-auth code but it provides us felxibility of fixing/improving spengo auth instead of trying to patch hadoop-auth which involves a bigger community and lot less inclination to change code.

https://github.com/apache/hadoop/tree/trunk/hadoop-common-project/hadoop-auth/src/main/java/org/apache/hadoop/security/authentication/server

If you see the hadoop-auth module, there isn't whole lot of activity going on or there are frequent changes in the project. So we will be able to monitor and pick changes that are needed. This shouldn't be a huge concern at this point and current module works for us for the most part.

satishd commented:
@harshach default authorizer implementation([PR-517](https://github.com/hortonworks/streamline/pull/517)) is done in streamline which we may want to have it in registry because of the reasons mentioned in your last [comment](https://github.com/hortonworks/registry/pull/52#issuecomment-288479317). So, we may also need common-authorization module to be pushed to registry like other common modules.

priyank5485 commented:
@satishd @hmcl @harshach @HeartSaVioR 

I went through the discussions above and it seems the consensus and the takeaway is to rename this module as per Satish's suggestion and putting it under a different group id as per Harsha's suggestion. We will need to revisit the authorizer code and probably moving it as well as suggested by Satish. If i am missing anything else here please let me know

harshach commented:
@satishd unlike authentication, authorizer will be a specific implementation as you are authorizing on resources in that project. I haven't gone through in details with Arun's implementation if thats a generic implementation we can definitely port it.

satishd commented:
@arunmahadevan Basic authorizer framework is independent of the resources which authorization can be applied. That can be extracted and reused in both streamline and registry. I am not advocating these modules should be in registry but it is good to share common modules. Discussed in earlier comments on how we would like to keep those common modules for now.
What do you think?

arunmahadevan commented:
@satishd the authorizer framework itself can be independent and re-used in registry, however a few things needs to be fixed and cleaned up.

1. Naming `StreamlineAuthorizer` -> `Authorizer`, mvn module `streamline-authorizer` -> `authorizer`
2. Some user facing interfaces are added in streamline-sdk. If we want it to be re-used, it should be within the authorizer module itself.
3. The underly tables are added to the streamline create and delete sql scripts. Maybe this needs to be separated out.

My preference is to do the above clean up and keep it in streamline for now and have it well tested. Whenever registry modules wants to use it, just include the maven dependency and fix any pending stuff to make it really independent. Later we can move it wherever we want. 

Overall I think we keep the modules wherever its most used/has the most dependency currently, and try to make it independent as and when required but refer in other project via maven dependencies and once it matures, we move it to the right place (if required)

hmcl commented on schema-registry-auth/src/main/java/com/hortonworks/registries/auth/client/AuthenticatedURL.java:
```diff
@@ -0,0 +1,278 @@
+/**
+ * Copyright 2016 Hortonworks.
```
@priyank5485 This one didn't get removed

priyank5485 replied:
@hmcl I have not pushed that fix yet. Its on my local branch. I will in a bit

hmcl commented on webservice/src/main/java/com/hortonworks/registries/webservice/RegistryApplication.java:
```diff
@@ -69,6 +72,21 @@ public void run(RegistryConfiguration registryConfiguration, Environment environ
             enableCORS(environment);
         }
+
+        List<ServletFilterConfiguration> servletFilterConfigurations = registryConfiguration.getServletFilters();
```
@priyank5485 let's put the code in lines 76-89 in their own method and call it something along the lines **addServletFilters** or **registerServletFilters**, following the pattern of small cohesive methods already existing in this **run** method. It will make the code much cleaner, modular, and easier to read.

priyank5485 replied:
@hmcl Will do

priyank5485 commented on common-auth/SECURITY.md:
```diff
@@ -0,0 +1,65 @@
+# Running registry and streamline web-services securely
+
+This module is intended to be used by registry and streamline web-services so that they can enable http client authentication via SPNEGO. Currently supported 
```
@hmcl I did not get the part about any registry dependent web-services. Should we really add that?

priyank5485 commented on common-auth/SECURITY.md:
```diff
@@ -0,0 +1,65 @@
+This module is intended to be used by registry and streamline web-services so that they can enable http client authentication via SPNEGO. Currently supported 
+authentication mechanism is Kerberos. The code for this module has been borrowed from hadoop-auth(2.7.3) module in hadoop project and changed a little. The 
+reasons for doing so are to avoid having a dependency on hadoop-auth module which brings in some other modules, avoid conflicts with other versions of 
+hadoop-auth module and having more control over the changes needed in future. Some text for this document has been borrowed from SECURITY.md of Apache Storm  
```
I don't mind removing that but it loses a lot of information that was suggested by others to be included in this doc. Lets keep it if you don't mind

hmcl replied:
Well, it says the same exact thing succinctly.

priyank5485 commented on common-auth/SECURITY.md:
```diff
@@ -0,0 +1,65 @@
+By default, registry and streamline web-services are running without authentication enabled and anyone can access the web-services from ui/client as far as they
+know the url and can access the web-server from the client machine. To enable authentication with the client, webservice needs to add a servlet filter from 
+this module. The webservice module will need to declare a dependency on this module. One way of adding a servlet filter in code is as follows. 
+
```
I am not sure if i understand correctly what the above means and if we should add that. May be i am missing something here.

hmcl replied:
This paragraph is very confusing to me. My understanding is: if a web service wants to use authentication, and that web service declares registry as a dependency, then the code snippet below illustrates how to do create a servlet filter. That's what I meant to say with my suggestion.If you want you can replace dependent with declares registry as dependency, or you can amend your original paragraph to be more clear.

priyank5485 replied:
This paragraph is just saying that if a webservice needs to use authentication they can use this module. It does not need to declare any dependency on registry top level maven module. Just the auth module here. The code snippet illustrates how you can add a add a servlet filter(which is why it says One way of adding.....) and shows how its done in registry web-service project. Code to add servlet filter is not part of this module. And they can add servlet filters in other ways too. Is it still unclear? Do you still want any modification to the paragraph?

priyank5485 commented on common-auth/SECURITY.md:
```diff
@@ -0,0 +1,65 @@
+ServletFilterConfiguration is a java object representing any servlet filter used in schema registry. However the general idea is that one needs to
+add com.hortonworks.registries.auth.server.AuthenticationFilter for enabling authentication 
+
+The configuration for that filter in yaml is represented in params property as follows.
```
I think its better to keep it. To give the reader a gist of what params is

hmcl replied:
oh, I know now what you mean. I would write it as:

The filter configuration is passed using the params property in they YAML file, as follows:

priyank5485 replied:
Will change

priyank5485 commented on common-auth/SECURITY.md:
```diff
@@ -0,0 +1,65 @@
+created as part of Kerberos setup
+
+Once configured, you must do `kinit` on client side using the principal of the user before accessing the web-service via the browser or some other client. 
+This principal also needs to be created first during Kerberos setup
```
thats for HTTP principal. This is for user principal. We should keep this

hmcl replied:
OK, but since I didn't understand that when I read it, perhaps we should try to make the text a bit more clear.

priyank5485 replied:
Any suggestions for text?

priyank5485 commented on common-auth/SECURITY.md:
````diff
@@ -0,0 +1,65 @@
+
+Here's an example of accessing web-service after the setup above:
+```bash
+curl  -i --negotiate -u:anyUser  -b ~/cookiejar.txt -c ~/cookiejar.txt  http://<web-service-host>:<port>/api/v1/
````
cookie jar is the file where cookies are stored when server first respnds with the cookie and we are passing the same back to the server in the request. I think this is more for man curl. We can avoid explaing how curl works here

hmcl replied:
ok, but if it was confusing to me, won't it be confusing to users reading this doc ?

priyank5485 replied:
It could be. How do you want me to word it? We described above what needs to be done on web server side and client side from Kerberos point of view and then we are just giving an example of a client aka curl. Even the settings change in browser is not clear and confusing. I did not understand when i read it in the first place. I think its out of the scope trying to explain documentation about browser settings or curl

priyank5485 commented on common-auth/SECURITY.md:
```diff
@@ -0,0 +1,65 @@
+3. IE: Configure trusted websites to include "web-service-hostname" and allow negotiation for that website
+
+**Caution**: In AD MIT Kerberos setup, the key size is bigger than the default UI jetty server request header size. If using that, make sure you set 
+header buffer bytes to 65536
```
In the request header

hmcl replied:
the text implies that the header buffer is to be set in the jetty server, which according to what you say is not the case. The text should be clear

priyank5485 replied:
What is the text you prefer? I will update it to that

hmcl replied:
**Caution**: In AD MIT Kerberos setup, the key size is bigger than the default UI jetty server request header size. If using MIT Kerberos, make sure you set the **(HTTP ?)** request header buffer bytes to 65536.

priyank5485 replied:
Will change.

priyank5485 commented:
@hmcl I have addressed most of your review comments on SECURITY.md. Put questions or replies where i was not sure. Definitely agree on common base module.

priyank5485 commented:
@hmcl pushed one more fix with a couple of changes

harshach commented:
+1
